src/cascadia/TerminalConnection/AzureConnection.cpp

```diff
@@ -0,0 +1,889 @@
+    {
+        // Initialize the client
+        http_client pollingClient(_loginUri);
+
+        // Continuously send a poll request until the user authenticates
+        const auto body = L"grant_type=device_code&resource=" + _wantedResource + L"&client_id=" + AzureClientID + L"&code=" + deviceCode;
+        const auto requestUri = L"common/oauth2/token";
+        json::value responseJson;
+        for (int count = 0; count < expiresIn / pollInterval; count++)
+        {
+            // User might close the tab while we wait for them to authenticate, this case handles that
+            if (_closing.load())
+            {
+                throw "Tab closed.";
+            }
+            http_request pollRequest(L"POST");
+            pollRequest.set_request_uri(requestUri);
+            pollRequest.set_body(body, L"application/x-www-form-urlencoded");
+
+            responseJson = _RequestHelper(pollingClient, pollRequest);
+
+            if (responseJson.has_field(L"error"))
+            {
+                Sleep(pollInterval * 1000); // Sleep takes arguments in milliseconds
+                continue; // Still waiting for authentication
+            }
+            else
+            {
+                _outputHandlers(winrt::to_hstring("Authenticated.\r\n"));
+                break; // Authentication is done, break from loop
+            }
+        }
+        if (responseJson.has_field(L"error"))
+        {
+            throw "Time out.";
+        }
+        return responseJson;
+    }
+
+    // Method description:
+    // - helper function to acquire the user's Azure tenants
+    // Return value:
+    // - the response which contains a list of the user's Azure tenants
+    json::value AzureConnection::_GetTenants()
+    {
+        // Initialize the client
+        http_client tenantClient(_resourceUri);
+
+        // Initialize the request
+        http_request tenantRequest(L"GET");
+        tenantRequest.set_request_uri(L"tenants?api-version=2018-01-01");
+        _HeaderHelper(tenantRequest);
+
+        // Send the request and return the response as a json value
+        return _RequestHelper(tenantClient, tenantRequest);
+    }
+
+    // Method description:
+    // - helper function to refresh the access/refresh tokens
+    // Return value:
+    // - the response with the new tokens
+    json::value AzureConnection::_RefreshTokens()
+    {
+        // Initialize the client
+        http_client refreshClient(_loginUri);
+
+        // Initialize the request
+        http_request refreshRequest(L"POST");
+        refreshRequest.set_request_uri(_tenantID + L"/oauth2/token");
+        const auto body = L"client_id=" + AzureClientID + L"&resource=" + _wantedResource + L"&grant_type=refresh_token" + L"&refresh_token=" + _refreshToken;
+        refreshRequest.set_body(body, L"application/x-www-form-urlencoded");
+        refreshRequest.headers().add(L"User-Agent", resource);

```

Comment:
We may need a better user agent than `Terminal`; perhaps we can get it from the package version (like App.cpp does?)